### PR TITLE
Reduce the amount of slicing in GetMemory/GetSpan/Advance

### PR DESCRIPTION
**Pull request by davidfowl**
- Avoid setting End on the writing segment until commit or until
we're moving to the next segment.
- This ends up making the pipe a little bigger but makes writing faster.

PS: This is another take on what I was trying to do with https://github.com/dotnet/corefx/pull/33948 (inspired by https://github.com/aspnet/AspNetCore/pull/7432)

This is the benchmark https://github.com/aspnet/AspNetCore/blob/c1ce7b9a7764f7ee742da7abb29c1b498f1e2594/src/Servers/Kestrel/perf/Kestrel.Performance/PipeThroughputBenchmark.cs but using the default pool instead of the Kestrel memory pool.

## Released version
```
                  Method |     Mean |     Error |    StdDev |   Median |      Op/s | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
------------------------ |---------:|----------:|----------:|---------:|----------:|------------:|------------:|------------:|--------------------:|
 ParseLiveAspNetTwoTasks | 1.002 us | 0.0815 us | 0.2379 us | 1.077 us | 998,447.5 |           - |           - |           - |                 1 B |
   ParseLiveAspNetInline | 1.113 us | 0.0839 us | 0.2448 us | 1.060 us | 898,460.3 |           - |           - |           - |                24 B |
```

## Before changes

```
                  Method |       Mean |     Error |   StdDev |        Op/s | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
------------------------ |-----------:|----------:|---------:|------------:|------------:|------------:|------------:|--------------------:|
 ParseLiveAspNetTwoTasks |   951.5 ns |  90.71 ns | 263.2 ns | 1,050,924.4 |           - |           - |           - |                 1 B |
   ParseLiveAspNetInline | 1,060.2 ns | 109.24 ns | 315.2 ns |   943,226.1 |           - |           - |           - |                   - |
```

## After changes
```
                  Method |     Mean |    Error |   StdDev |        Op/s | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
------------------------ |---------:|---------:|---------:|------------:|------------:|------------:|------------:|--------------------:|
 ParseLiveAspNetTwoTasks | 894.1 ns | 87.71 ns | 255.8 ns | 1,118,447.7 |           - |           - |           - |                 1 B |
   ParseLiveAspNetInline | 991.6 ns | 67.47 ns | 190.3 ns | 1,008,508.4 |           - |           - |           - |                   - |
```

### CODE REVIEW

**Comment by benaadams** [comment]
I think it can go further; checking

File: src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs [hunk 1]
```diff
@@ -168,16 +166,13 @@ internal Span<byte> GetSpan(int sizeHint)
                 AllocateWriteHeadUnsynchronized(sizeHint);
```
Comment (benaadams) [inline src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs]: Can put this in 
```csharp
if (_writingMemory.Length < Math.Max(1, sizeHint))
```
Would also need to move
```
_operationState.BeginWrite();
```
from `AllocateWriteHeadUnsynchronized` here (and to `GetMemory`)

File: src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs [hunk 2]
```diff
@@ -144,11 +146,7 @@ internal Memory<byte> GetMemory(int sizeHint)
...
```
Comment (benaadams) [inline src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs]: Same as below

File: src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs (hunk 1 above)
Comment (davidfowl) [reply src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs]: I'm not yet convinced this is right. _writingMemory would need to be updated with _writingHead.

Comment (benaadams) [reply src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs]: Would need lock free handlining on `_operationState` so probably out of scope (for reset in `AdvanceReader`)

Comment (benaadams) [reply src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs]: k something like?
```csharp
if (!_operationState.IsWritingActive ||
    _writingMemory.Length < Math.Max(1, sizeHint))
{
    lock (_sync)
    {
        AllocateWriteHeadUnsynchronized(sizeHint);
    }
}

return _writingMemory.Span;
```

Comment (benaadams) [reply src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs]: Which would work for the `BufferWriter` pattern where it runs out of space and does
```csharp
_output.Advance(_buffered);
_buffered = 0;
_span = _output.GetSpan(count);
```

**Comment by benaadams** [comment]
Looks good
image

**Comment by benaadams** [comment]
`corefx-ci` failure due to timeout

File: src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs [hunk 3]
```diff
@@ -295,16 +303,14 @@ internal void Advance(int bytesWritten)
...
+                if (bytesWritten > _writingMemory.Length)
                 {
                     ThrowHelper.ThrowInvalidOperationException_AdvancingPastBufferSize();
                 }
 
-                // if bytesWritten is zero, these do nothing
-                _writingHead.End += bytesWritten;
                 _currentWriteLength += bytesWritten;
+                _buffered += bytesWritten;
+                _writingMemory = _writingMemory.Slice(bytesWritten);
```
Comment (pakrym) [inline src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs]: Can you help me understand why this change helped. Seems like Slice just moved out from .End setter to here.

Comment (benaadams) [reply src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs]: Slice in `.End` doesn't add value to the writer as its a `ReadOnlyMemory` so doing that it still needs to be resliced on `GetMemory`

This now Slices the `Memory` then so its always read for the writer, then Slices via `.End` for the reader on `.Commit`

Comment (davidfowl) [reply src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs]: There are 2 sets of slicing that need to happen:
 1. Changing the memory for the reader - `AvaliableMemory.Slice(0, bytesWritten)`
 2. Changing the memory for the writer - `AvailableMemory.Slice(bytesWritten)`

Today every call to GetMemory/GetSpan does a slice to give expose the writable space and Advance updates the memory available to the reader. It means when you GetMemory/GetSpan and Advance we update both pieces of memory all of the time. This change makes it so that we update the appropriate side when necessary (on Flush and when we run out of space in the existing segment). Other performance comes from the fact that we're not going through the segment every time to get the writable memory since it's just a field.

Comment (pakrym) [reply src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs]: Why don't we just have 2 properties on Segment then? One for writing one for reading pointing to the same memory?

Comment (pakrym) [reply src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs]: > Other performance comes from the fact that we're not going through the segment every time to get the writable memory since it's just a field.

Is this proven? How expensive is the indirection?

Comment (benaadams) [reply src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs]: > Why don't we just have 2 properties on Segment then?

Then you'll bloat the segment size of which there are many for the writing head which there is one?

Comment (benaadams) [reply src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs]: Specifically by 16 bytes
```
Type layout for 'Memory`1'
Size: 16 bytes. Paddings: 0 bytes (%0 of empty space)
|=================================|
|   0-7: Object _object (8 bytes) |
|---------------------------------|
|  8-11: Int32 _index (4 bytes)   |
|---------------------------------|
| 12-15: Int32 _length (4 bytes)  |
|=================================|
```

Comment (pakrym) [reply src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs]: Now I understand, thank you all.

Comment (benaadams) [reply src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs]: Motivation behind this change is @davidfowl thought doing this upstream was an abomination https://github.com/aspnet/AspNetCore/pull/7432 😄
